### PR TITLE
Safetensors serialization by default

### PR DESCRIPTION
**Pull request by LysandreJik**
This PR aims to do one thing but is larger than expected. I'm happy to break it down into smaller PRs if it helps for reviewing.

This PR aims to switch safe serialization to `True` by default for `torch` models. In doing so, it revealed a few bugs in the existing implementation and `safetensors` support that this PR fixes.

Additionally, support for `safetensors` for Flax models is added so that models saved from PyTorch after merging this PR can be used in both TensorFlow and Flax, and for models saved from TensorFlow/Flax to be loaded in PyTorch models.

The following should be worked on shortly to enable switching to safetensors by default for TensorFlow and Flax as well:
- There is no support for sharded weights in TensorFlow
- There is no support for sharded weights in Flax

Additionally, I'll contribute some documentation making the following clear:
- TensorFlow models can load models in safetensors saved from PyTorch and TensorFlow, but it cannot load them from Flax. This can be eventually worked on; meanwhile, I'll write this in the docs with workarounds to get models saved in Flax to work in TensorFlow for those interested.
- Same, but for Flax models loaded from TensorFlow

Thanks, @Rocketknight1, for the help on TensorFlow's side.

### CODE REVIEW

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -3063,6 +3057,29 @@ def from_pretrained(
...
+                logger.info("A TensorFlow safetensors file is being loaded in a PyTorch model.")
+            elif metadata.get("format") == "flax":
+                from_flax = True
+                logger.info("A Flax safetensors file is being loaded in a PyTorch model.")
+            else:
+                raise ValueError(
+                    f"Incompatible safetensors file. File metadata is not ['pt', 'tf', 'flax'] but {metadata.get('format')}"
+                )
+
+        from_pt = not (from_tf | from_flax)
```
Comment (LysandreJik) [inline src/transformers/modeling_utils.py:3081]: this is necessary to enable loading `safetensors` files saved from TensorFlow/Jax into PyTorch models

File: src/transformers/modeling_flax_utils.py [hunk 2]
```diff
@@ -1029,6 +1104,7 @@ def save_pretrained(
         push_to_hub=False,
         max_shard_size="10GB",
         token: Optional[Union[str, bool]] = None,
+        safe_serialization: bool = False,
```
Comment (LysandreJik) [inline src/transformers/modeling_flax_utils.py:1107]: This is kept to `False` by default for now; switching it to `True` is possible in the short term, but we would need to support sharded weights first.

File: tests/models/auto/test_modeling_tf_auto.py [hunk 3]
```diff
@@ -211,6 +211,8 @@ def test_from_pretrained_with_tuple_values(self):
         config = copy.deepcopy(model.config)
         config.architectures = ["FunnelBaseModel"]
         model = TFAutoModel.from_config(config)
+        model.build()
```
Comment (Narsil) [inline tests/models/auto/test_modeling_tf_auto.py:214]: Is this related to the change ? Was it done elsewhere before (Cannot find it)

File: tests/test_modeling_utils.py [hunk 4]
```diff
@@ -1070,6 +1080,84 @@ def test_generation_config_is_loaded_with_model(self):
...
+            new_model = BertModel.from_pretrained(tmp_dir)
+
+        for p1, p2 in zip(model.parameters(), new_model.parameters()):
+            self.assertTrue(torch.equal(p1, p2))
+
+    @require_safetensors
+    @require_flax
+    def test_safetensors_torch_from_flax_sharded(self):
+        # Not implemented yet
+        pass
```
Comment (Narsil) [inline tests/test_modeling_utils.py:1135]: To remove ? (There's another occurrence)

**Review by Narsil — APPROVED** [review]
Very nice ! 

Not *that* bad for such a big change.

For the testing part I see loading in PT from PT/TF/Flax, but not the other ways 

TF -TF
TF - Flax
TF - Pt
Flax - Flax
Flax - TF
Flax - Pt.

From you initial comment I understand it's not possible, but it's not entirely clear for me as to why (you mention sharded weights, is it the only restriction? If yes, from what I read it should be okay-ish to be able to at least load for those, no ?)

File: src/transformers/modeling_flax_pytorch_utils.py [hunk 5]
```diff
@@ -319,11 +344,15 @@ def load_flax_checkpoint_in_pytorch_model(model, flax_checkpoint_path):
...
 
     # load flax weight dict
-    with open(flax_checkpoint_path, "rb") as state_f:
-        try:
-            flax_state_dict = from_bytes(flax_cls, state_f.read())
-        except UnpicklingError:
-            raise EnvironmentError(f"Unable to convert {flax_checkpoint_path} to Flax deserializable object. ")
+    if flax_checkpoint_path.endswith(".safetensors"):
+        flax_state_dict = safe_load_file(flax_checkpoint_path)
+        flax_state_dict = unflatten_dict({tuple(k.split(".")): v for k, v in flax_state_dict.items()})
```
Comment (sanchit-gandhi) [inline src/transformers/modeling_flax_pytorch_utils.py:349]: Equivalent to:
```suggestion
        flax_state_dict = unflatten_dict(flax_state_dict, sep=".")
```

File: src/transformers/modeling_flax_utils.py [hunk 6]
```diff
@@ -422,6 +433,31 @@ def to_fp16(self, params: Union[Dict, FrozenDict], mask: Any = None):
         ```"""
         return self._cast_floating_to(params, jnp.float16, mask)
 
+    @classmethod
+    def load_flax_weights(cls, resolved_archive_file):
+        try:
+            if resolved_archive_file.endswith(".safetensors"):
+                state = safe_load_file(resolved_archive_file)
+                state = unflatten_dict({tuple(k.split(".")): v for k, v in state.items()})
```
Comment (sanchit-gandhi) [inline src/transformers/modeling_flax_utils.py:441]:
```suggestion
                state = unflatten_dict(state, sep=".")
```

File: src/transformers/modeling_flax_utils.py [hunk 7]
```diff
@@ -687,7 +723,19 @@ def from_pretrained(
...
+                ):
+                    # Load from a safetensors checkpoint
+                    archive_file = os.path.join(pretrained_model_name_or_path, SAFE_WEIGHTS_NAME)
+                elif is_safetensors_available() and os.path.isfile(
+                    os.path.join(pretrained_model_name_or_path, SAFE_WEIGHTS_INDEX_NAME)
+                ):
+                    # Load from a sharded safetensors checkpoint
+                    archive_file = os.path.join(pretrained_model_name_or_path, SAFE_WEIGHTS_INDEX_NAME)
+                    is_sharded = True
+                    raise NotImplementedError("Support for sharded checkpoints using safetensors is coming soon!")
```
Comment (sanchit-gandhi) [inline src/transformers/modeling_flax_utils.py:737]: If a repository has both:
1. Sharded safetensors weights
2. Sharded Flax weights

Then we currently cannot load the sharded safetensors weights into Flax (which is fine), but we shouldn't let this block the user loading the sharded Flax weights for now?

Should we move the check for sharded safetensors weights to *after* our check for sharded Flax weights? This way we don't block the user loading the latter if available.

File: src/transformers/modeling_flax_utils.py [hunk 8]
```diff
@@ -722,7 +770,13 @@ def from_pretrained(
                 filename = pretrained_model_name_or_path
                 resolved_archive_file = download_url(pretrained_model_name_or_path)
             else:
-                filename = WEIGHTS_NAME if from_pt else FLAX_WEIGHTS_NAME
+                if from_pt:
```
Comment (sanchit-gandhi) [inline src/transformers/modeling_flax_utils.py:773]: If we set `from_pt=True` but safetensors weights are available, then we still use the safetensors weights no? So should we re-order this if/else logic to:
```python
                if is_safetensors_available():
                    filename = SAFE_WEIGHTS_NAME
                elif from_pt:
                    filename = WEIGHTS_NAME
                else:
                    filename = FLAX_WEIGHTS_NAME
```
This then matches the order we set when we check the model path

File: src/transformers/modeling_flax_utils.py [hunk 9]
```diff
@@ -742,6 +796,22 @@ def from_pretrained(
...
                     # Since we set _raise_exceptions_for_missing_entries=False, we don't get an expection but a None
                     # result when internet is up, the repo and revision exist, but the file does not.
+                    if resolved_archive_file is None and filename == SAFE_WEIGHTS_NAME:
+                        # Maybe the checkpoint is sharded, we try to grab the index name in this case.
+                        resolved_archive_file = cached_file(
+                            pretrained_model_name_or_path, SAFE_WEIGHTS_INDEX_NAME, **cached_file_kwargs
+                        )
+                        if resolved_archive_file is not None:
+                            is_sharded = True
+                            raise NotImplementedError(
```
Comment (sanchit-gandhi) [inline src/transformers/modeling_flax_utils.py:806]: Same here - can we just let the user download the sharded Flax weights in this case? Instead of erroring out when loading the sharded safetensors ones?

File: src/transformers/modeling_flax_utils.py [hunk 10]
```diff
@@ -820,31 +891,35 @@ def from_pretrained(
...
+            safetensors_from_pt = safetensors_metadata.get("format") == "pt"
+            safetensors_metadata.get("format") == "tf"
+
         # init random models
         model = cls(config, *model_args, _do_init=_do_init, **model_kwargs)
 
-        if from_pt:
+        if safetensors_from_pt:
+            state_dict = {}
+            with safe_open(resolved_archive_file, framework="pt") as f:
```
Comment (sanchit-gandhi) [inline src/transformers/modeling_flax_utils.py:911]: (nit) Should this loading of the torch tensors go into the function `load_pytorch_state_dict_in_flax_state_dict`? This would be consistent with `load_pytorch_checkpoint_in_flax_state_dict` where we load the PT state dict inside the function: https://github.com/huggingface/transformers/blob/84724efd101af52ed3d6af878e41ff8fd651a9cc/src/transformers/modeling_flax_pytorch_utils.py#L59

File: src/transformers/modeling_flax_utils.py [hunk 11]
```diff
@@ -820,31 +891,35 @@ def from_pretrained(
...
+        if filename == SAFE_WEIGHTS_NAME:
+            with safe_open(resolved_archive_file, framework="flax") as f:
+                safetensors_metadata = f.metadata()
+            if safetensors_metadata is None or safetensors_metadata.get("format") not in ["pt", "tf", "flax"]:
+                raise OSError(
+                    f"The safetensors archive passed at {resolved_archive_file} does not contain the valid metadata."
+                    " Make sure you save your model with the `save_pretrained` method."
+                )
+            safetensors_from_pt = safetensors_metadata.get("format") == "pt"
+            safetensors_metadata.get("format") == "tf"
```
Comment (sanchit-gandhi) [inline src/transformers/modeling_flax_utils.py:904]: Is this line here by accident?

File: src/transformers/modeling_flax_utils.py (hunk 2 above)
Comment (sanchit-gandhi) [reply src/transformers/modeling_flax_utils.py:1107]: See https://github.com/huggingface/transformers/blob/84724efd101af52ed3d6af878e41ff8fd651a9cc/src/transformers/modeling_flax_utils.py#L93

File: tests/test_modeling_flax_utils.py [hunk 12]
```diff
@@ -184,3 +185,72 @@ def test_model_from_pretrained_hub_subfolder_sharded(self):
...
+            self.assertTrue(os.path.isfile(os.path.join(tmp_dir, SAFE_WEIGHTS_NAME)))
+            self.assertFalse(os.path.isfile(os.path.join(tmp_dir, FLAX_WEIGHTS_NAME)))
+
+            new_model = FlaxBertModel.from_pretrained(tmp_dir)
+
+        self.assertTrue(check_models_equal(model, new_model))
+
+    @require_flax
+    @require_torch
+    def test_safetensors_save_and_load_pt_to_tf(self):
```
Comment (sanchit-gandhi) [inline tests/test_modeling_flax_utils.py:205]:
```suggestion
    def test_safetensors_save_and_load_pt_to_flax(self):
```

File: tests/test_modeling_flax_utils.py [hunk 13]
```diff
@@ -184,3 +185,72 @@ def test_model_from_pretrained_hub_subfolder_sharded(self):
         model = FlaxBertModel.from_pretrained(model_id, subfolder=subfolder)
 
         self.assertIsNotNone(model)
+
+    @require_safetensors
```
Comment (sanchit-gandhi) [inline tests/test_modeling_flax_utils.py:189]: Nice tests!

File: src/transformers/modeling_flax_utils.py [hunk 14]
```diff
@@ -1101,24 +1179,32 @@ def save_pretrained(
...
                 os.remove(full_filename)
 
         if index is None:
-            with open(output_model_file, "wb") as f:
+            if safe_serialization:
                 params = params if params is not None else self.params
-                model_bytes = to_bytes(params)
-                f.write(model_bytes)
+                flat_dict = flatten_dict(params)
+                state_dict = {".".join(k): v for k, v in flat_dict.items()}
```
Comment (sanchit-gandhi) [inline src/transformers/modeling_flax_utils.py:1201]: Same for flatten dict!
```suggestion
                flat_dict = flatten_dict(params, sep=",")
```

**Review by sanchit-gandhi — APPROVED** [review]
Thanks for working on this @LysandreJik!

File: src/transformers/modeling_flax_pytorch_utils.py (hunk 5 above)
Comment (LysandreJik) [reply src/transformers/modeling_flax_pytorch_utils.py:349]: wooo this is a very cool method. Thanks, that's a nice tip!

File: tests/models/auto/test_modeling_tf_auto.py (hunk 3 above)
Comment (LysandreJik) [reply tests/models/auto/test_modeling_tf_auto.py:214]: Yes, this is related to the change! It wasn't done before and therefore was working purely out of luck.
After instantiating a TensorFlow model with `from_config`, one should build it as it otherwise has no weights attached to it.

Saving weights through the current mechanism (h5) works even with no weights attached to the model; however, that isn't the case for `safetensors` which raises an error. This caused issues when passing to `safe_serialization=True`, so I fixed the test to work correctly.

The specific error was:
```
Error
Traceback (most recent call last):
  File "/transformers/tests/models/auto/test_modeling_tf_auto.py", line 220, in test_from_pretrained_with_tuple_values
    model = TFAutoModel.from_pretrained(tmp_dir)
  File "/transformers/src/transformers/models/auto/auto_factory.py", line 565, in from_pretrained
    return model_class.from_pretrained(
  File "/transformers/src/transformers/modeling_tf_utils.py", line 2877, in from_pretrained
    with safe_open(resolved_archive_file, framework="tf") as f:
safetensors_rust.SafetensorError: Error while deserializing header: InvalidHeaderDeserialization
```

File: src/transformers/modeling_flax_utils.py (hunk 7 above)
Comment (LysandreJik) [reply src/transformers/modeling_flax_utils.py:737]: Very nice catch! That's also the case for TensorFlow, actually. Modifying this and adding tests

File: src/transformers/modeling_flax_utils.py (hunk 11 above)
Comment (LysandreJik) [reply src/transformers/modeling_flax_utils.py:904]: Ah yes, typo from trying to enable loading from TensorFlow safetensors :grin: Thanks for the catch!

File: src/transformers/modeling_flax_utils.py (hunk 7 above)
Comment (LysandreJik) [reply src/transformers/modeling_flax_utils.py]: Addressed in https://github.com/huggingface/transformers/pull/27064/commits/1bd29e03c09284653e1d8ae4a8ed426299b99083

File: src/transformers/modeling_flax_utils.py (hunk 8 above)
Comment (LysandreJik) [reply src/transformers/modeling_flax_utils.py]: Correct!

Addressed in https://github.com/huggingface/transformers/pull/27064/commits/1bd29e03c09284653e1d8ae4a8ed426299b99083

File: src/transformers/modeling_flax_utils.py (hunk 9 above)
Comment (LysandreJik) [reply src/transformers/modeling_flax_utils.py]: Correct!

Addressed in https://github.com/huggingface/transformers/pull/27064/commits/1bd29e03c09284653e1d8ae4a8ed426299b99083

**Comment by LysandreJik** [comment]
@Narsil, this is what is currently supported and not supported:

- TF -TF - Supported, tested here: https://github.com/huggingface/transformers/blob/a79600fc26d57a9262ce069434e3bc0ea6b8ef47/tests/test_modeling_tf_utils.py#L501-L509
- TF - Flax - **Not supported**
- TF - Pt - Supported, tested here: https://github.com/huggingface/transformers/blob/a79600fc26d57a9262ce069434e3bc0ea6b8ef47/tests/test_modeling_tf_utils.py#L511-L522
- Flax - Flax - Supported, tested here: https://github.com/huggingface/transformers/blob/a79600fc26d57a9262ce069434e3bc0ea6b8ef47/tests/test_modeling_flax_utils.py#L236-L244
- Flax - TF - **Not supported**
- Flax - Pt - Supported, tested here: https://github.com/huggingface/transformers/blob/a79600fc26d57a9262ce069434e3bc0ea6b8ef47/tests/test_modeling_flax_utils.py#L246-L256

> From you initial comment I understand it's not possible, but it's not entirely clear for me as to why (you mention sharded weights, is it the only restriction? If yes, from what I read it should be okay-ish to be able to at least load for those, no ?)

I mention this in the PR description:

> TensorFlow models can load models in safetensors saved from PyTorch and TensorFlow, but it cannot load them from Flax. This can be eventually worked on; meanwhile, I'll write this in the docs with workarounds to get models saved in Flax to work in TensorFlow for those interested.

It should be pretty straightforward to enable it, but I suspect extremely little usage for a TF <> Flax conversion where no PyTorch conversion exists. I'm planning to add this to the documentation and IMO we can work on it afterwards if there are requests.

File: tests/test_modeling_utils.py (hunk 4 above)
Comment (LysandreJik) [reply tests/test_modeling_utils.py]: Addressed in https://github.com/huggingface/transformers/pull/27064/commits/c979b854defed0eae80ce1e97330851e9d5ddbd3

File: src/transformers/modeling_flax_utils.py (hunk 8 above)
Comment (LysandreJik) [reply src/transformers/modeling_flax_utils.py]: Actually thinking more about it, I think that when using `from_pt=True` we'll want to continue forcing the load of the `pytorch_model.bin`. The safetensors file will be used as default in all other cases

File: src/transformers/modeling_tf_utils.py [hunk 15]
```diff
@@ -2366,8 +2368,8 @@ def save_pretrained(
 
             create_pr (`bool`, *optional*, defaults to `False`):
                 Whether or not to create a PR with the uploaded files or directly commit.
-            safe_serialization (`bool`, *optional*, defaults to `False`):
-                Whether to save the model using `safetensors` or the traditional PyTorch way (that uses `pickle`).
+            safe_serialization (`bool`, *optional*, defaults to `False``):
```
Comment (amyeroberts) [inline src/transformers/modeling_tf_utils.py:2371]:
```suggestion
            safe_serialization (`bool`, *optional*, defaults to `False`):
```

File: tests/test_modeling_flax_utils.py [hunk 16]
```diff
@@ -184,3 +185,88 @@ def test_model_from_pretrained_hub_subfolder_sharded(self):
...
+            )
+
+            # This should not raise even if there are two types of sharded weights
+            FlaxBertModel.from_pretrained(path)
+
+    @require_safetensors
+    def test_safetensors_flax_from_sharded_msgpack_with_sharded_safetensors_hub(self):
+        # This should not raise even if there are two types of sharded weights
+        # This should discard the safetensors weights in favor of the msgpack sharded weights
+        FlaxBertModel.from_pretrained("hf-internal-testing/tiny-bert-flax-safetensors-msgpack-sharded")
```
Comment (amyeroberts) [inline tests/test_modeling_flax_utils.py:272]: Just to make sure I've understood correctly: TF and Flax models can't load sharded weights from safetensors. So, if this passes, we know the model has successfully loaded the msgpack sharded weights?

File: tests/test_modeling_tf_utils.py [hunk 17]
```diff
@@ -496,6 +497,44 @@ def test_safetensors_load_from_hub(self):
...
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            path = snapshot_download("hf-internal-testing/tiny-bert-tf-safetensors-h5-sharded", cache_dir=tmp_dir)
+
+            # This should not raise even if there are two types of sharded weights
+            TFBertModel.from_pretrained(path)
+
+    @require_safetensors
+    def test_safetensors_flax_from_sharded_msgpack_with_sharded_safetensors_hub(self):
+        # This should not raise even if there are two types of sharded weights
+        # This should discard the safetensors weights in favor of the msgpack sharded weights
```
Comment (amyeroberts) [inline tests/test_modeling_tf_utils.py:535]: I think?

```suggestion
        # This should discard the safetensors weights in favor of the .h5 sharded weights
```

File: tests/test_modeling_tf_utils.py [hunk 18]
```diff
@@ -496,6 +497,44 @@ def test_safetensors_load_from_hub(self):
...
+    @require_safetensors
+    def test_safetensors_tf_from_sharded_h5_with_sharded_safetensors_local(self):
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            path = snapshot_download("hf-internal-testing/tiny-bert-tf-safetensors-h5-sharded", cache_dir=tmp_dir)
+
+            # This should not raise even if there are two types of sharded weights
+            TFBertModel.from_pretrained(path)
+
+    @require_safetensors
+    def test_safetensors_flax_from_sharded_msgpack_with_sharded_safetensors_hub(self):
```
Comment (amyeroberts) [inline tests/test_modeling_tf_utils.py:533]: I think? Doesn't appear to involve flax in this test

```suggestion
    def test_safetensors_tf_from_sharded_h5_with_sharded_safetensors_hub(self):
```

File: src/transformers/modeling_flax_utils.py [hunk 19]
```diff
@@ -35,10 +35,14 @@
 from .configuration_utils import PretrainedConfig
 from .dynamic_module_utils import custom_object_save
 from .generation import FlaxGenerationMixin, GenerationConfig
-from .modeling_flax_pytorch_utils import load_pytorch_checkpoint_in_flax_state_dict
+from .modeling_flax_pytorch_utils import (
+    load_pytorch_checkpoint_in_flax_state_dict,
+)
```
Comment (amyeroberts) [inline src/transformers/modeling_flax_utils.py:40]: ultra-ultra nit 

```suggestion
from .modeling_flax_pytorch_utils import load_pytorch_checkpoint_in_flax_state_dict
```

File: src/transformers/modeling_flax_utils.py [hunk 20]
```diff
@@ -820,31 +886,27 @@ def from_pretrained(
...
+                    f"The safetensors archive passed at {resolved_archive_file} does not contain the valid metadata."
+                    " Make sure you save your model with the `save_pretrained` method."
+                )
+            safetensors_from_pt = safetensors_metadata.get("format") == "pt"
+
         # init random models
         model = cls(config, *model_args, _do_init=_do_init, **model_kwargs)
 
-        if from_pt:
+        if from_pt or safetensors_from_pt:
```
Comment (amyeroberts) [inline src/transformers/modeling_flax_utils.py:903]: Exposing my lack of knowledge about safe tensors here: if `safetensors_from_pt` is `True` here then is the reason we do `load_pytorch_checkpoint_in_flax_state_dict(model, resolved_archive_file, is_sharded)` because the serialized weights are in "pytorch format" and therefore can't be loaded using `cls.load_flax_weights`?

File: src/transformers/modeling_tf_utils.py [hunk 21]
```diff
@@ -986,15 +988,15 @@ def load_tf_weights_from_safetensors(model, resolved_archive_file, ignore_mismat
     # Read the safetensors file
     with safe_open(resolved_archive_file, framework="tf") as safetensors_archive:
         mismatched_layers = []
-        weight_names = [format_weight_name(w.name, _prefix=_prefix) for w in model.weights]
+        weight_names = [strip_model_name_and_prefix(w.name, _prefix=_prefix) for w in model.weights]
```
Comment (amyeroberts) [inline src/transformers/modeling_tf_utils.py:991]: As: 
* The previous function `format_weight_name` added `_prefix`, whereas the new function `strip_model_name_and_prefix` removes `_prefix` if it's present. 
* `weight_names` are compared to those in `safetensors_archive`

Can we load in previously saved safetensors (before this PR) into our TF models?

File: src/transformers/modeling_tf_utils.py [hunk 22]
```diff
@@ -2782,21 +2784,12 @@ def from_pretrained(
                     # Since we set _raise_exceptions_for_missing_entries=False, we don't get an exception but a None
                     # result when internet is up, the repo and revision exist, but the file does not.
                     if resolved_archive_file is None and filename == SAFE_WEIGHTS_NAME:
-                        # Maybe the checkpoint is sharded, we try to grab the index name in this case.
+                        # Did not find the safetensors file, let's fallback to Flax.
```
Comment (amyeroberts) [inline src/transformers/modeling_tf_utils.py:2787]:
```suggestion
                        # Did not find the safetensors file, let's fallback to TF.
```

**Review by amyeroberts — APPROVED** [review]
Very nice! 🔥 

Just some small nits and Qs for my own understanding

File: tests/test_modeling_flax_utils.py (hunk 16 above)
Comment (LysandreJik) [reply tests/test_modeling_flax_utils.py:272]: Yes that's exactly right! This was raised by Sanchit as a previous version of the implementation priorized safetensors, realized they were sharded, and errored-out; but if sharded msgpack are also in the repo, we would want to load these first

File: src/transformers/modeling_flax_utils.py (hunk 20 above)
Comment (LysandreJik) [reply src/transformers/modeling_flax_utils.py:901]: Yes, that's correct! This way we call `load_pytorch_checkpoint_in_flax_state_dict` with the safetensors file, and that method checks if the file ends with `.safetensors` to load it the pytorch-way

File: src/transformers/modeling_tf_utils.py (hunk 21 above)
Comment (LysandreJik) [reply src/transformers/modeling_tf_utils.py:991]: cc @Rocketknight1

Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py:991]: @amyeroberts I believe the previous code was just completely incorrect! Essentially, the relevant workflow (saving TF composite encoder-decoder models as safetensors and then reloading the checkpoint in TF) was not being tested, and actually didn't work because of the name prefix bug.

As such, I don't think there's a backwards compatibility issue here, because previous checkpoints weren't working at all. My suspicion is that not many people were saving encoder-decoder models in TF, and not many TF users were saving safetensors, and so the intersection of that venn diagram was tiny enough that no-one noticed the bug for a long time!

Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py:991]: Also, just to clarify: `_prefix` is almost always `None` when this function is called, in which case the behaviour is unchanged after this bugfix. `_prefix` is only defined when loading composite models like `EncoderDecoder`, which is the workflow that was broken before this.

**Comment by LysandreJik** [comment]
I will proceed to merge this and write a small explanatory doc tomorrow. I would like for the slow tests to run on this before the release.

File: src/transformers/modeling_tf_utils.py (hunk 21 above)
Comment (amyeroberts) [reply src/transformers/modeling_tf_utils.py:991]: Thanks for explaining!

File: tests/test_modeling_flax_utils.py (hunk 16 above)
Comment (amyeroberts) [reply tests/test_modeling_flax_utils.py:272]: Thanks for explaining!

File: src/transformers/modeling_flax_utils.py (hunk 20 above)
Comment (amyeroberts) [reply src/transformers/modeling_flax_utils.py:901]: Thanks for explaining!

**Comment by Narsil** [comment]
Awesome ! Thanks a LOT for this.
